### ISSUE TITLE
npm bundled with node 0.12.2 basically doesn't work at all 

### ISSUE THREAD

**Issue by tjwebb** [npm/npm#8100]
First, I can't publish stuff: https://github.com/npm/npm/issues/8099. 

I can't install either. I get a bunch of warnings, npm hangs, and then after a couple minutes, it dies.

```
$ npm install
npm WARN peerDependencies The peer dependency stream-browserify@* included from bl will no
npm WARN peerDependencies longer be automatically installed to fulfill the peerDependency 
npm WARN peerDependencies in npm 3+. Your application will need to depend on it explicitly.
npm WARN optional dep failed, continuing fsevents@0.3.5
npm WARN optional dep failed, continuing fsevents@0.3.5
```

Here it sits and does nothing for awhile, then finally...

```
npm ERR! Linux 3.13.0-49-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v0.12.2
npm ERR! npm  v2.7.4
npm ERR! code ECONNRESET
npm ERR! errno ECONNRESET
npm ERR! syscall read

npm ERR! network read ECONNRESET
npm ERR! network This is most likely not a problem with npm itself
npm ERR! network and is related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network 
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'
```

and I'm not behind a freaking proxy. If I just downgrade back to node 0.10.35, everything works fine.

**Comment by tjwebb** [comment]
Also if I upgrade to npm 2.8.3, everything works again.

**Comment by smikes** [comment]
You are encountering the interaction between two bugs, one in node >0.11 (and all iojs) http-keepalive, and one in `npm-registry-client`.  Any version of `npm` >= 2.8.2 has a fix for the `npm-registry-client` bug, which works around the bug in http-keepalive.

Main issue is here - https://github.com/npm/npm/issues/7699

When the next `node` 0.12 comes out, we hope to have a fixed `npm` in there.

**Comment by tjwebb** [comment]
I love how new versions of npm and node are bundled together, yet they apparently are tested independently. fail.

**Comment by smikes** [comment]
You're absolutely right— it's really a problem that the community has gotten fragmented in this way.  In this case, we were only able to see the bug after lots of people reported problems with npm install on node 0.12. 

The http-keepalive client code in all versions of node >= 0.11 exhibits this race condition, and this includes all io.js versions too.

At present we are just working around the bug- not triggering it by not writing to the socket in our GET request. We don't even have a proper issue report / repro case on the iojs or node trackers yet. :-(

**Comment by tjwebb** [comment]
>  it's really a problem that the community has gotten fragmented in this way.

Good thing npm is supporting iojs over node, a fork that resulted from someone becoming disgruntled because they didn't receive an immediate response to their bug report over Thanksgiving. You know, the US national holiday. Lucky for npm, we all *have* to use it. There's not another decent option quite yet.

**Comment by othiym23** [comment]
The wording "supporting iojs over node" isn't an accurate way of describing the attitude or behavior the project towards Node.js _or_ io.js. Node.js 0.12.3 was just released, and includes `npm@2.9.1`, which I worked with @misterdjules to land. I've also regularly been submitting new versions of npm to the io.js team. We've been testing npm under both Node.js and io.js, and have been more or less agnostic about which runtime you use.

A relatively small number of users got bitten by the issue @smikes described in Node.js 0.12.2, although I'm not sure the problem you described and the bug he described were the same thing. Node 0.12.2 / `npm@2.7.4` were tested together before the release, by both @misterdjules and me, and for whatever reason, we didn't encounter this issue. I'll be the first to admit that there's room for improvement in npm's test suite (I would like to see it rewritten nearly entirely eventually), but we ran the tests we had.

Now that Node.js and io.js are converging under the aegis of the Node Foundation, the fork will become a moot point, so I'm a little confused as to why you're bringing it up now. I'm optimistic that the work being done towards an LTS version of Node will be fruitful, and am spending some time thinking about how npm could best support that project. Either way, it has never been the policy of npm to play favorites when it comes to the fork, and I'm not sure what it does for anybody to claim otherwise.

**Comment by tjwebb** [comment]
> isn't an accurate way of describing the attitude or behavior of the project towards Node.js or io.js.

I've read and listened to (e.g. javascript jabber) quite a few of Isaac's thoughts on the matter. If I've misinterpreted what he's saying and doing, or if his actions don't represent the official policy of npm, then I apologize.

> so I'm a little confused as to why you're bringing it up now.

Because I was still using node 0.12.2 (and whichever npm version 2.7.broken) this morning when I updated some of my modules, prompting some of my users to complain. So yea, I was already irritated when linked to this issue, and I saw @smike's response.

**Comment by smikes** [comment]
The worst part is, I was trying to be sympathetic, but ended up making things worse.  Sorry.

**Comment by isaacs** [comment]
> I've read and listened to (e.g. javascript jabber) quite a few of Isaac's thoughts on the matter. If I've misinterpreted what he's saying and doing, or if his actions don't represent the official policy of npm, then I apologize.

My actions do represent official policy of npm.  That's what the CEO job entails :)

My statements and actions have been that we will continue to support Node, io.js, and any other sufficiently modern node-like runtime (from v0.8 and up).  Saying that io.js happened only because of one disgruntled developer, or concluding that I've given up on 97% of our users because the core team is reworking our process under a new brand, is dropping rather a lot of context and history.

Bugs happen in software and we fix them as we find them.  There is no way around this.  I understand it is very frustrating to encounter a bug in "latest" shipping software that is allegedly fixed by the maintainers already.  Even with 100% test coverage, it would have been tricky to find this bug, since it's a combination of weird behavior between two software systems outside of npm.

We landed a workaround, added a new test, and moved forward.  The underlying issue has been fixed in io.js, I'm not sure if it's fixed in Node 0.12.  And anyway, that didn't help you, because you weren't using the npm version with the fix.  We aren't in charge of shipping joyent/node, so we couldn't get you the fix.

This is why regular coordinated releases are a necessary part of shipping stable software, part of the problem that io.js's process innovations are designed to correct.  Those process improvements will be continued under the name "Node.js" in The Node Foundation starting with version 3.

**Comment by misterdjules** [comment]
> The underlying issue has been fixed in io.js, I'm not sure if it's fixed in Node 0.12.

I would be happy to know what the underlying issue was, so that we can fix it in node too if it hasn't been done yet.

**Comment by misterdjules** [comment]
@isaacs @othiym23 Could you please point me to "The underlying issue has been fixed in io.js", I would like to make sure it's been fixed in node v0.12 too.

Thank you!

**Comment by othiym23** [comment]
@misterdjules I'm fairly certain @smikes was referring to the net unref race condition that @indutny fixed for Node.js 0.12.2. Although if not, he can correct me. ;)
